verifiable-db/src/query/merkle_path.rs

```diff
@@ -291,37 +349,344 @@ where
+
+    let (mut predecessor_info, mut successor_info) = if with_neighbors {
+        // we need to initialize predecessor and successor data
+        let end_node_info = end_node_info.as_ref().unwrap();
+        // the predecessor of end_node is an ancestor of end_node iff end_node has no left child
+        let is_predecessor_in_path = b.not(end_node_info.left_child_exists);
+        let zero_u256 = b.zero_u256();
+        // Initialize value of predecessor node of end_node to a dummy value if the predecessor node
+        // will be found in the path; otherwise, the predecessor_value is the maximum value in
+        // the subtree rooted in the left child of end_node
+        let predecessor_value = b.select_u256(
+            is_predecessor_in_path,
+            &zero_u256,
+            &end_node_info.left_child_info.max,
+        );
+        // the predecessor value is already found if end_node has a left child
+        let predecessor_found = end_node_info.left_child_exists;
+        // Initialize predecessor node hash to a dummy value
+        let predecessor_hash = b.constant_hash(*empty_poseidon_hash());
+        // build predecessor info
+        let predecessor_info = NeighborInfoTarget {
+            is_found: predecessor_found,
+            is_in_path: is_predecessor_in_path,
+            value: predecessor_value,
+            hash: predecessor_hash,
+        };
+
+        // the successor of end_node is an ancestor of end_node iff end_node has no right child
+        let is_successor_in_path = b.not(end_node_info.right_child_exists);
+        // Initialize value of successor node of end_node to a dummy value if the successor node
+        // will be found in the path; otherwise, successor_value is the minimum value in
+        // the subtree rooted in the right child of end_node
+        let successor_value = b.select_u256(
+            is_successor_in_path,
+            &zero_u256,
+            &end_node_info.right_child_info.min,
+        );
+        // the successor value is already found if end_node has a right child
+        let successor_found = end_node_info.right_child_exists;
+        // Initialize successor node hash to a dummy value
+        let successor_hash = b.constant_hash(*empty_poseidon_hash());
+        // build successor info
+        let successor_info = NeighborInfoTarget {
+            is_found: successor_found,
+            is_in_path: is_successor_in_path,
+            value: successor_value,
+            hash: successor_hash,
+        };
+        (Some(predecessor_info), Some(successor_info))
+    } else {
+        (None, None)
+    };
+
+    let mut final_hash = end_node;
+    for i in 0..MAX_DEPTH - 1 {
+        let rest = node_min[i]
+            .to_targets()
+            .into_iter()
+            .chain(node_max[i].to_targets())
+            .chain(once(index_id))
+            .chain(node_value[i].to_targets())
+            .chain(embedded_tree_hash[i].to_targets())
+            .collect_vec();
```

Comment:
Could we use the node input target struct which already defines the sequence of elements to hash ?

Reply:
I am not sure I fully understand what you mean: you are suggesting to add a method to `NodeInfoTarget` that will compute `rest` basically? If so, I am not sure how to name this method, computing `rest` inside of the whole hash is an optimization but it's not a "well-defined" operation. Wdyt?